ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs

```diff
@@ -98,46 +98,55 @@ type family HasResponder (appType :: AppType) :: Bool where
 --   serving downstream peers using server side of each protocol and getting
 --   updates from upstream peers using client side of each of the protocols.
 --
+-- * the monadic action which wraps the callbacks will be run only once by the
+--   multiplexing layer.  This makes it safe for sharing a common mutable state
+--   between various protocol.
```

Comment:
When will it be run? Why can't the programmer run it before setting up the mux/server stuff?

Reply:
Clarification: it will be run only once _per peer_. Correct?

Reply:
Indeed, once per connection. Probably much better type is this one:
```
data MuxApplication (appType :: AppType) ptcl m bytes a where
  MuxInitiaroApplication
    :: forall s appType ptcl m bytes a.
        m s
    -> (s -> ptcl -> Channel m bytes -> m a)
    -> MuxApplication appType ptcl m bytes a
...
```
What I need is a shared state between all the `ptcl` threads, which is constructed before starting all the mux threads.

Reply:
This is needed because there's synchronisation between chain sync and block fetch clients in the consensus layer  (chain sync client has to start after block fetch).

Reply:
> chain sync client has to start after block fetch

What's the reason for it? Seems like something that ought to be fixed.

Reply:
I'd imagine (admitting that I know very little of the actual implementation) that block fetch and chain sync parts would synchronize by way of some shared STM state

Reply:
It's explained here: https://github.com/input-output-hk/ouroboros-network/blob/f3d6569bbed21628252c6200f6c63309acb886e4/ouroboros-consensus/src/Ouroboros/Consensus/Node.hs#L300